text/0104-string-parameterization-output-spec.md

```diff
@@ -0,0 +1,264 @@
+```
+
+A successful parameterizer will ingest these span descriptions, and determine that `"CA"`, `"US"`, and `"AU"` were probably a parameter inserted dynamically by the application depending on some condition. The three query descriptions are examples of this parameterized query:
+
+```sql
+SELECT * FROM cities WHERE county = ?;
+```
+
+where `?` is a country code parameter that varies from query to query.
+
+This spec outlines standards for the output of a parameterizer, including but not limited to the desired string formatting, payload structure, and character escaping. The spec is organized by data type. Parameterization methods and infrastructure are outside the scope of this spec.
+
+## Motivation
+
+The main goal of parameterization is to remove dynamic values from strings, in order to group “the same” strings together. “The same” loosely means that the strings represent the same underlying resource. For example, the strings `GET /users/17` and `GET /users/781` might represent accessing the web application route `/users/:id`. The strings `SCARD org:18:active` and `SCARD org:98:active` represent access to the same kind of Redis resource, and so on. When looking at application data (traces, URLs, spans, etc.) it is most useful to look at parameterized values, because they represent aggregate behaviour. In the routing example, the `/users/:id` route is a useful entity to examine, while the unparameterized URLs are not. Parameterization enables aggregation, and aggregation makes it possible to store, query, and compare data easily.
+
+Sentry performs parameterization in many different part of the application. The SDKs, Relay, and Ingest are three examples. There is no agreement between these parts beyond conventions. As a result, the output of parameterization is inconsistent.
+
+Here are three ways different SDKs might parameterize a SQL query:
+
+- `SELECT * FROM countries WHERE id = ?`
+- `SELECT * FROM countries WHERE id = %s`
+- `SELECT * FROM countries WHERE id = $1`
+
+As another example, the SDKs might pass the query `SELECT * FROM country WHERE code IN (?, ?)` to Ingest, which will further parameterize that as `SELECT * FROM country WHERE code IN (?)`. The string `(?, ?)` is technically only partially parameterized from the system's perspective.
+
+This has important downstream effects on user experience. A good parameterizer will correctly group similar strings together in a way that represents the original intent. If parameterization is successful, users will see groups that make sense to them. Good parameterization will also reduce cardinality, and therefore reduce data storage costs. A bad parameterizer will fill the users’s lists of transactions and spans with noisy similar descriptions that drown out the signal.
+
+This spec documents the desired output of a parameterizer. This makes it possible for different parameterizers to agree, conform to a standard, share code, and create consistency in the system.
```

Comment:
I am not sure why this consistency is even desired. Ensuring that all SQL uses `?` as parameter can feel "foreign" to a user familiar with a particular database. In the case of C-style formatting, `%(named_parameter)` vs `%(named_parameter)s` also has different meaning. The spec also locks in named parameters in some cases to be formatted with curly braces, which only feels native in some languages (rust, python)

Is there something specific we want to do in the UI with those parameters, such as syntax-highlight them? It might be possible to use _meta protocol from Relay here, though it is a very complex solution. Basically what it does is to allow you to annotate ranges of characters of a string with custom data:

```
{
   "myvalue": "/foo/{bar}/baz",
   "_meta": {"myvalue": {"": "rem": [["@parametrizer","s",6,10]]
}
```

Reply:
Expanding this a bit more: this goal tries to set a standard across all organizations, while specific organizations may follow a single pattern and don't have different formats. In the latter case, we'd be reformatting their queries for no real benefit. I don't think we should have a different approach per organization but approach this on a language basis, as Markus suggests (we can always have a fallback in Relay).

Reply:
> Ensuring that all SQL uses ? as parameter can feel "foreign" to a user familiar with a particular database.

So far, the product folks don't think this will be a problem, but I'm curious about this. Do you know of a SQL-like database that uses a different character?

> The spec also locks in named parameters in some cases to be formatted with curly braces, which only feels native in some languages (rust, python)

I waffled on this a lot. I think curlies are familiar to almost all developers. Python, Ruby, JavaScript, Rust, PHP all use some kind of curly-like syntax for interpolation though they're not all the same.

> Is there something specific we want to do in the UI with those parameters, such as syntax-highlight them?

This is a great point, thanks for raising. I will think about this more. No, the spec is not concerned with ideas like syntax highlighting, but the `_meta` protocol is interesting!

Reply:
> I don't think we should have a different approach per organization but approach this on a language basis

I think I know what you mean, but I'm not sure, could you elaborate?